### PR TITLE
MINOR Speed up DependentPages by shifting existence check to table join

### PR DESCRIPTION
**Pull request by maxime-rainville**
This PR updates `SiteTree::BackLinkTracking` to use some convoluted inner joins to avoid having to do individual query on each result to check for its existence.

It also alter queries to add the `DependentLinkType` alias field in the select statement which seems to give us a slight performance improvement over looping over all the results and adding it manually.

# Parent issue
* https://github.com/silverstripe/silverstripe-versioned/issues/189

### CODE REVIEW

**Comment by robbieaverill** [comment]
Please consider #2327 when reviewing this. I don't think we should continue any longer using an ArrayList here, it's simply bad for performance

**Comment by maxime-rainville** [comment]
Totally agree that this is just patching over the fundamental problem that `SiteTreeLink` is poorly thought.

I think this PR should be seen as a stopgap to allow us to ship key project. Rethinking Dependent Pages to work properly will probably require us to deprecated `SiteTreeLink` and build something else.

Looks like my change broke something. I'll have a look and try to fix it.

**Comment by maxime-rainville** [comment]
Fixed it for real now. My change to `BackLinkTracking()` was returning a list of `SiteTreeLink` rather than returning the list of Parent Object like the original.

File: code/Model/SiteTree.php [hunk 1]
```diff
@@ -1727,19 +1727,42 @@ public function getLiveURLSegment()
...
+
+        // Get the list of back links classes
+        $parentClasses = $this->BackLinks()->exclude(['ParentClass' => null])->columnUnique('ParentClass');
+
+        // Get list of sitreTreelink and join them to the their parent class to make sure we don't get orphan records.
+        foreach ($parentClasses as $parentClass) {
+            $joinClause = sprintf(
+                "\"%s\".\"ParentID\"=\"%s\".\"ID\"",
+                $siteTreelinkTable,
+                $linkTable,
```
Comment (robbieaverill) [inline code/Model/SiteTree.php]:
```suggestion
```

**Pull request merged**